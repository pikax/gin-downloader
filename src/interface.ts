import {MangaFilter} from "./filter";
import {FilterStatus, Type} from "./enum";
import {OptionsWithUrl, Options, Response, OptionsWithUri} from "request";
import {
    IFilterSource,
    IGenreSite,
    IMangaConfig,
    IMangaFilter,
    IMangaParser,
    IMangaRequest,
    IMangaVisitor
} from "./manga/interface";
import {MangaHereConfig} from "./manga/mangahere/config";
import {mangahereLogger} from "./manga/mangahere/logger";
import {MangaHereVisitor} from "./manga/mangahere/visitor";
import {MangahereParser} from "./manga/mangahere/parser";
import {MangaHereFilter} from "./manga/mangahere/filter";
import {MangaHereGenre} from "./manga/mangahere/genre";
import {ILogger} from "./util/logger";
import {IMangaRequestFactory} from "../build/manga/interface";
import {Partial} from "rollup-plugin-typescript2/dist/partial";


export interface Source {
    src: string;
}

export interface IReadOnlyMangaOld {
    readonly name: string;
    readonly status?: string;
    readonly mature?: boolean;
    readonly image?: string;

    readonly info: ILazy<Promise<MangaInfo>>;
    readonly chapters: ILazy<Promise<ChapterCollection>>;
}


export interface IManga {
    name: string;
    status?: string;
    mature?: boolean;
    image?: string;

}


export interface IMangaExtended extends IReadOnlyManga {
    readonly info: Promise<MangaInfo>;
    readonly chapters: Promise<IChapterExtended[]>;
}

export interface IReadOnlyManga extends Readonly<IManga> {

}


export interface IChapter {
    chap_number?: string;
    volume?: string;
    name?: string;

    language?: string;
    scanlator?: string;
    dateAdded?: string;

    licensed?: boolean;
}

export interface IChapterExtended extends IChapter {

    readonly images: Promise<ImageCollection>;

    readonly manga: Promise<IMangaExtended>;
}

export interface Synonym {
    title: string;
    language: string;
}

export interface Info {
    image: string;
    title: string;

    synonyms?: Synonym[];
    released?: string;
    authors?: string[];
    artists?: string[];
    genres?: string[];
    synopsis?: string;
    status?: string;

    type?: Type;
    licensed?: boolean;
    mature?: boolean;


    // site specific
    ranked?: string;
    rating?: string;
    scanlators?: string[];
    similar?: string[];
    views?: string;
}

export interface GinImage extends Source {
    name: string;
}

export interface MangaResults {
    results: IManga[];
    page: number;
    total: number;
}


export interface ILazy<T> {
    readonly value: T;
}


export interface MangaInfo {
    image: string;
    title: string;

    synonyms?: Synonym[];
    released?: string;
    authors?: string[];
    artists?: string[];
    genres?: string[];
    synopsis?: string;

    status?: FilterStatus;
    ranked?: string;
    rating?: string;
    scanlators?: string[];
    similarmanga?: string[];
    direction?: string;

    views?: string;
    type?: string;

    licensed?: boolean;
}


export interface IRequestStrategy {
    request(options: OptionsWithUri): Promise<Response>;
}


export interface IImageFactory {
    create();
}

export interface IMangaFactory {
    create(src: string, manga: IManga);
}

export interface IChapterFactory {
    create();
}


export interface IMangaObject {

    info(): Promise<Info>;

    chapters(): Promise<IChapter[]>;

    images(chapter): Promise<ImageCollection>;
}


export interface IChapterResolver {
    chapters(src): Promise<ChapterSource[]>;
}

export interface IImageResolver {
    images(src): Promise<ImageCollection>;
}

export interface IMangaInfoResolver {
    info(src): Promise<MangaInfo>;
}


export interface IMangaConfigDependency {
    config: IMangaConfig;
}

export interface IMangaLoggerDependency {
    logger: ILogger;
}

export interface IMangaGenreDependency {
    genre: IGenreSite;
}

export interface IMangaFilterDependency {
    filter: IFilterSource;
}

export interface IMangaParserDependency {
    parser: IMangaParser;
}

export interface IMangaVisitorDependency {
    visitor: IMangaVisitor;
}

export interface IMangaRequestFactoryDependency {
    requestFactory: IMangaRequestFactory;
}


export interface IMangaDependencies extends IMangaConfigDependency, IMangaLoggerDependency, IMangaGenreDependency, IMangaFilterDependency, IMangaParserDependency, IMangaVisitorDependency, IMangaRequestFactoryDependency {

}

export type RequestFactoryMangaDependencies = Partial<IMangaDependencies> & IMangaRequestFactoryDependency;

export interface IMangaBuilder {
    build(di: RequestFactoryMangaDependencies): IMangaDependencies;
}


export type GinImagePromise = ILazy<Promise<GinImage>>;
export type ChapterCollection = IChapter[];
export type MangaCollection = IManga[];
export type ImageCollection = GinImagePromise[];
export type InfoChapter = Info & { chapters: ChapterCollection };


export type ChapterSource = IChapter & { src: string };
