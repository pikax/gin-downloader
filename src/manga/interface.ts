import {
    IChapter,
    MangaCollection,
    MangaInfo,
    ChapterSource
} from "../interface";
import {filter, FilteredResults, MangaSource} from "../filter";
import FilterSupport = filter.FilterSupport;

import {Genre} from "../enum";
import {OptionsWithUri} from "request";

import * as cheerio from "cheerio";


export interface IMangaLogin {
    login(user: string, password: string): Promise<boolean>;
}


export interface IMangaSite {
    mangas(): Promise<MangaCollection>;

    // info(): Promise<MangaInfo>;
    // chapters(): Promise<ChapterCollection>;
}


export interface IMangaFilter {
    filter(filter: FilterSupport): Promise<FilteredResults>;
}


export interface IMangaLatest {
    latest(): Promise<MangaCollection>;
}

export interface IMangaVisitor {
    latest(): IterableIterator<{ href: string, page: number; total: number }>;

    mangas(): IterableIterator<{ href: string, page: number; total: number }>;
}

export interface IMangaRequest {
    readonly uri: string;

    readonly html: string;

    readonly $: CheerioStatic;
}

export interface IMangaRequestFactory {
    request(options: OptionsWithUri): Promise<IMangaRequest>;
}


export class MangaRequestResult implements IMangaRequest {
    private _$: CheerioStatic;

    get uri() {
        return this._uri;
    }

    get html() {
        return this._html;
    }

    get $() {
        return this._$ || (this._$ = cheerio.load(this.html));
    }

    constructor(private _uri: string, private _html: string) {
    }
}


export interface IMangaParser {
    mangas(mangaRequest: IMangaRequest): IterableIterator<MangaSource>;

    // latest chapters
    latest(mangaRequest: IMangaRequest): IterableIterator<IChapter & { src: string }>;

    // info manga
    info(mangaRequest: IMangaRequest): MangaInfo;

    // chapters
    chapters(mangaRequest: IMangaRequest): IterableIterator<ChapterSource>;


    // image urls
    imagesPaths(mangaRequest: IMangaRequest): IterableIterator<{ name: string; src: string }>;

    // single image
    image(mangaRequest: IMangaRequest): string;


    // returns the current filter page
    filterPage(mangaRequest: IMangaRequest): { page: number; total: number };
}


export interface IMangaConfig {
    readonly name: string;
    readonly site: string;

    readonly mangasUrl: string;
    readonly latestUrl: string;
}


// resolves the url and params a website
export interface IFilterSource {
    process(filter: FilterSupport): IFilterSourceResult;
}

export interface IFilterSourceResult {
    src: string;
    params?: any;
}

export enum MangaSite {
    MangaHere = "mangahere",
}


export interface IGenreSite {
    toSiteGenre(genre: Genre): string;

    fromSiteGenre(genre: string): Genre;

    isSupported(genre: Genre): boolean;

    supported(): Genre[];
}


//
// export class SuperMangaSite {
//
//     constructor(private _config: IMangaConfig, private _requestFactory: IMangaRequestFactory, private _parser: IMangaParser, private _visitor: IMangaVisitor) {
//
//     }
//
//     mangas(): Promise<IMangaExtended[]> {
//         return null;
//     }
//
//     async latest(): Promise<IChapterExtended> {
//         // const self = this;
//
//         // return co(function* () {
//         //
//         //     for (const c of self._visitor.latest()) {
//         //         yield *
//         //     }
//         //
//         // });
//
//
//         const it = await this.latestProc(this._config.latestUrl);
//
//         it.map(x => new SuperMangaChapter())
//
//
//     }
//
//     async latestProc(src: string) {
//         const result = await this._requestFactory.request({
//             uri: src
//         });
//
//         return [...this._parser.latest(result)];
//     }
//
//     /*
//         async* latest(): Promise<Iterable<IChapterExtended>> {
//             const latestUrl = this._config.latestUrl;
//
//             const result = await this._requestFactory.request({
//                 uri: latestUrl
//             });
//
//
//             for (const c of this._parser.latest(result)) {
//
//                 // yield new
//             }
//
//
//         }*/
//
//
//     async allImagesResolver(chapterSrc: string): Promise<ILazy<Promise<ImageSource>>[]> {
//         let href = url.resolve(this._config.site, chapterSrc).toString();
//         let result = await this._requestFactory.request({uri: href});
//
//         const paths = this._parser.imagesPaths(result);
//         paths.next(); // skip first, because we requested that already
//
//         let src = this._parser.image(result);
//
//         const promises = [
//             new Lazy<Promise<ImageSource>>(() => Promise.resolve({
//                 name: url.parse(src).pathname.split("/").reverse()[0],
//                 src
//             }))
//         ];
//
//
//         const promiseRequest = async (s): Promise<ImageSource> => {
//             const h = url.resolve(href, s).toString();
//             const result = await this._requestFactory.request({uri: h})
//             const src = this._parser.image(result);
//
//             return {
//                 name: url.parse(src).pathname.split("/").reverse()[0],
//                 src
//             };
//         };
//
//
//         for (const it of paths) {
//             promises.push(new Lazy<Promise<ImageSource>>(() => promiseRequest(it.src)));
//         }
//
//
//         return promises;
//     }
//
//
//     * allImagesResolver2(chapterSrc: string) {
//         const it = this._imagesResolver(chapterSrc);
//
//         it.next(); // first request
//
//         yield* it;
//     }
//
//     _request(href) {
//         return this._requestFactory.request({uri: href});
//     }
//
//
//     * _imagesResolver(chapterSrc: string) {
//         let href = url.resolve(this._config.site, chapterSrc).toString();
//         let result = yield this._requestFactory.request({uri: href});
//
//         const paths = this._parser.imagesPaths(result);
//         paths.next(); // skip first, because we requested that already
//
//         let src = this._parser.image(result);
//
//
//         yield new Lazy<Promise<ImageSource>>(() => Promise.resolve({
//             name: url.parse(src).pathname.split("/").reverse()[0],
//             src
//         }))
//
//
//         const promiseRequest = async (s): Promise<ImageSource> => {
//             const h = url.resolve(href, s).toString();
//             const result = await this._requestFactory.request({uri: h})
//             const src = this._parser.image(result);
//
//             return {
//                 name: url.parse(src).pathname.split("/").reverse()[0],
//                 src
//             };
//         };
//
//
//         for (const it of paths) {
//             yield new Lazy<Promise<ImageSource>>(() => promiseRequest(it.src));
//         }
//
//     }
//
//
//     //
//     // async* imagesResolver2(chapterSrc: string): Iterable<ImageSource> {
//     //     let href = url.resolve(this._config.site, chapterSrc).toString();
//     //
//     //     let result = await this._requestFactory.request({uri: href});
//     //
//     //     const paths = this._parser.imagesPaths(result);
//     //     paths.next(); // skip first, because we requested that already
//     //
//     //
//     //     let src = this._parser.image(result);
//     //
//     //     yield {
//     //         name: url.parse(src).pathname.split("/").reverse()[0],
//     //         src
//     //     };
//     //
//     //
//     //     for (const p of paths) {
//     //         const h = url.resolve(href, p.src).toString();
//     //         result = await this._requestFactory.request({uri: h})
//     //         src = this._parser.image(result);
//     //
//     //         yield {
//     //             name: url.parse(src).pathname.split("/").reverse()[0],
//     //             src
//     //         };
//     //     }
//     //
//     // }
//
// }
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
