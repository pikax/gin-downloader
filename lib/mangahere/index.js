'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _parser = require('./parser');

var _parser2 = _interopRequireDefault(_parser);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _request = require('../common/request');

var _names = require('./names');

var _helper = require('../common/helper');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Created by rodriguesc on 03/03/2017.
 */
var debug = require('debug')('gin-downloader:mangahere');
var verbose = require('debug')('gin-downloader:mangahere:verbose');

var mangas = function mangas() {
  debug('getting mangas');
  return (0, _helper.getDoc)(_config2.default.mangas_url).then(_parser2.default.mangas).tap(function (x) {
    return debug('mangas: ' + x.length);
  });
};

var latest = function latest() {
  debug('getting latest');
  return (0, _helper.getDoc)(_config2.default.latest_url).then(_parser2.default.latest).tap(function (x) {
    return debug('found: ' + x.length);
  });
};

var info = function info(name) {
  debug('getting info for ' + name);

  var src = (0, _names.resolveUrl)(name);

  debug('getting info ' + src);
  return (0, _helper.getDoc)(src).then(_parser2.default.mangaInfo).tap(function (x) {
    return debug('info: %s', x);
  });
};

var chapters = function chapters(name) {
  debug('getting chapters ' + name);
  var src = (0, _names.resolveUrl)(name);

  return (0, _helper.getDoc)(src).then(_parser2.default.chapters).tap(function (x) {
    if (!x || x.length === 0) throw 'not found';
  }).catch(function (x) {
    throw new Error('Chapters not found', x);
  }).tap(function (x) {
    return debug('chapters: ' + x.length);
  });
};

var images = function images(name, chapter) {
  debug('getting images ' + name + ':' + chapter);

  return chapters(name).then(function (chaps) {
    return _lodash2.default.find(chaps, { number: name + ' ' + chapter });
  }).then(function (x) {
    if (!x) throw new Error('Manga: ' + name + ' chapter ' + chapter + ' doesn\'t exists.');
    return x.src;
  }).then(imagesByUrl);
};

var imagesByUrl = function imagesByUrl(uri) {
  debug('getting images for ' + uri);

  return imagesPaths(uri).tap(function (x) {
    return debug('images: ' + x.length);
  }).then(function (x) {
    return x.map(image);
  }).tap(function (x) {
    return debug('images resolved : ' + x);
  });
};

var imagesPaths = function imagesPaths(uri) {
  debug('getting image paths ' + uri);
  return (0, _helper.getDoc)(uri).then(_parser2.default.imagesPaths).tap(function (x) {
    return debug('found ' + x.length + ' images:\n' + x);
  });
};

var image = function image(uri) {

  debug('getting image from ' + uri);

  return (0, _request.getHtml)(uri).tap(function (x) {
    return verbose('html: %s', x);
  }).then(_parser2.default.image);
};

var resolve = function resolve(name, chapter) {
  debug('resolve ' + name + ':' + chapter);

  return images(name, chapter).tap(function (x) {
    return debug('found %o', x);
  });
};

exports.default = {
  config: _config2.default,

  mangas: mangas,
  info: info,
  chapters: chapters,
  images: images,
  imagesPaths: imagesPaths,
  resolve: resolve,
  latest: latest
};