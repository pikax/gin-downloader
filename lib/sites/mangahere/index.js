"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Created by rodriguesc on 03/03/2017.
 */
var mangasite_1 = require("../../common/mangasite");
var parser_1 = require("./parser");
var config_1 = require("./config");
var names_1 = require("./names");
var request_1 = require("../../common/request");
var filter_1 = require("./filter");
var MangaHere = (function (_super) {
    __extends(MangaHere, _super);
    function MangaHere() {
        return _super.call(this, config_1.config, new parser_1.Parser(), new names_1.Helper(), request_1.request) || this;
    }
    MangaHere.prototype.filter = function (filter) {
        return __awaiter(this, void 0, void 0, function () {
            var search, doc, mangas;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.debug("filter mangas with: %o", filter);
                        search = filter_1.processFilter(filter);
                        return [4 /*yield*/, this.request.getDoc(search.src + "?" + search.params)];
                    case 1:
                        doc = _a.sent();
                        return [4 /*yield*/, this.parser.filter(doc)];
                    case 2:
                        mangas = _a.sent();
                        this.debug("mangas: " + mangas.results.length);
                        return [2 /*return*/, mangas];
                }
            });
        });
    };
    return MangaHere;
}(mangasite_1.MangaSite));
exports.MangaHere = MangaHere;
exports.manga = new MangaHere();
exports.default = exports.manga;
//# sourceMappingURL=index.js.map