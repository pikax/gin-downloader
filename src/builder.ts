import {MangaHere, MangaHereDependencies} from "./manga/mangahere";
import {MangaHereBuilder} from "./manga/mangahere/builder";
import {RequestRetryStrategy} from "./strategies/retry_strategy";
import {ConcurrentQueueRequestFactory} from "./factories/concurrentQueueRequest";
import {DefaultResolverFactory} from "./factories/resolvers";

const Queue = require("promise-queue");

export type GinBuilderDependencies = {};

export class GinBuilder {
    private _mangahere: MangaHere;
    private _mangaHereBuilder: MangaHereBuilder;
    private _requestFactory: ConcurrentQueueRequestFactory;
    private _requestStrategy: RequestRetryStrategy;
    private _resolverFactory: DefaultResolverFactory;


    get mangahere() {
        return this._mangahere || (this._mangahere = this.buildMangaHere());
    }


    constructor(dependencies: GinBuilderDependencies = {}) {
        this._requestStrategy = new RequestRetryStrategy();
        this._resolverFactory = new DefaultResolverFactory();
        this._requestFactory = new ConcurrentQueueRequestFactory(this._requestStrategy, new Queue(3));

        this._mangaHereBuilder = new MangaHereBuilder();
    }


    private buildMangaHere() {
        return new MangaHere({requestFactory: this._requestFactory, resolverFactory: this._resolverFactory});
    }


}

