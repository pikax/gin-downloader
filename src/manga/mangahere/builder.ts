import {MangaHereFilter} from "./filter";
import {MangaHereGenre} from "./genre";
import {mangahereLogger} from "./logger";
import {MangaHereVisitor} from "./visitor";
import {MangaHereConfig} from "./config";
import {MangahereParser} from "./parser";
import {IMangaBuilder, IMangaDependencies, RequestFactoryMangaDependencies} from "../../interface";


export class MangaHereBuilder implements IMangaBuilder {
    build(di: RequestFactoryMangaDependencies): IMangaDependencies {
        const {requestFactory} = di;

        if (!requestFactory) {
            throw new Error("Please provide request factory");
        }

        const logger    = di.logger     || mangahereLogger;
        const genre     = di.genre      || new MangaHereGenre();
        const config    = di.config     || new MangaHereConfig();
        const filter    = di.filter     || new MangaHereFilter(genre);
        const parser    = di.parser     || new MangahereParser(logger, config, genre);
        const visitor   = di.visitor    || new MangaHereVisitor(requestFactory, config);

        return {
            requestFactory,
            logger,
            genre,
            config,
            filter,
            parser,
            visitor
        };
    }
}


