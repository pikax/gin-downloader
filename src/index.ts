/**
 * Created by david on 25/03/2017.
 */


import "./declarations";

import {MangaFox} from "./sites/mangafox";
import {MangaHere} from "./sites/mangahere";
import {MangaPanda} from "./sites/mangapanda";


export const gin = {
  MangaFox,
  MangaHere,
  MangaPanda
};

export default gin;

