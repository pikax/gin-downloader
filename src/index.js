/**
 * Created by david on 19/03/2017.
 */

import fs from 'fs';

import builder from './sites';


let sites = {};

const sitesDir = __dirname + '/sites/';
fs.readdirSync(sitesDir).forEach((file)=> {
  let fp = sitesDir + file;

  if( fs.statSync(fp).isFile())
    return; //nothing to do here

  let site = require(fp).default;
  let parser = require(fp + '/parser').default;
  let config = require(fp + '/config').default;


  let def = new builder(site,config,parser);

  let s = site;

  s.mangas = s.mangas || def.mangas.bind(def);

  site[config.name] = s;
});


modules.export = site;


