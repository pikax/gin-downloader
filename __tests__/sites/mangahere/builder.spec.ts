import {MangaHereBuilder} from "../../../src/manga/mangahere/builder";
import {
    RetryRequestFactory
} from "../../../src/manga/interface";
import {RequestRetryStrategy} from "../../../src/strategies/retry_strategy";
import {RequestFactoryMangaDependencies} from "../../../src/interface";

describe("MangaHere builder", function () {
    const requestFactory = new RetryRequestFactory(new RequestRetryStrategy());

    it("should fail if no requestFactory", () => {
        const func = () => new MangaHereBuilder().build({requestFactory: null});

        expect(func).toThrowErrorMatchingSnapshot();
    });


    it("should resolve dependencies", () => {
        const di = new MangaHereBuilder().build({requestFactory});

        expect(di.requestFactory).toBeDefined();
        expect(di.visitor).toBeDefined();
        expect(di.parser).toBeDefined();
        expect(di.filter).toBeDefined();
        expect(di.config).toBeDefined();
        expect(di.genre).toBeDefined();
        expect(di.logger).toBeDefined();
    });

    it("should keep passed dependencies", () => {
        const config = {} as any;
        const logger = {} as any;
        const genre = {} as any;
        const filter = {} as any;
        const parser = {} as any;
        const visitor = {} as any;

        const expectedDi: RequestFactoryMangaDependencies = {
            requestFactory,
            logger,
            config,
            genre,
            filter,
            parser,
            visitor
        };

        const di = new MangaHereBuilder().build(expectedDi);

        expect(di).toEqual(expectedDi);
    });

    it("should keep partial passed dependencies", () => {
        const logger = {} as any;
        const parser = {} as any;
        const visitor = {} as any;

        const expectedDi: RequestFactoryMangaDependencies = {
            requestFactory,
            logger,
            parser,
            visitor
        };

        const di = new MangaHereBuilder().build(expectedDi);

        expect(di).not.toEqual(expectedDi);


        expect(di.requestFactory);
        expect(di.requestFactory).toBe(expectedDi.requestFactory);
        expect(di.logger).toBe(expectedDi.logger);
        expect(di.parser).toBe(expectedDi.parser);
        expect(di.visitor).toBe(expectedDi.visitor);


        expect(di.config).toBeDefined();
        expect(di.config).not.toEqual(expectedDi.config);
    });


});

