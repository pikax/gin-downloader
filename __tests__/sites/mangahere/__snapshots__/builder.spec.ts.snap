// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaHere builder should fail if no requestFactory 1`] = `"Please provide request factory"`;
